File: app/src/main/java/com/vikingsen/cheesedemo/model/data/price/PriceRepository.java
package com.vikingsen.cheesedemo.model.data.price;

import java.util.concurrent.TimeUnit;

import javax.inject.Inject;
import javax.inject.Singleton;

import io.reactivex.Single;
import kotlin.Pair;

@Singleton
public class PriceRepository {

    private final PriceRemoteDataSource remoteDataSource;

    @Inject
    PriceRepository(PriceRemoteDataSource remoteDataSource) {
        this.remoteDataSource = remoteDataSource;
    }

    public Single<Price> getPrice(long cheeseId, boolean forceRefresh) {
        return Single.just(new Pair<>(cheeseId, forceRefresh))
                .flatMap(pair -> {

                    try {
                        TimeUnit.SECONDS.sleep(2);
                    } catch (Exception ignore) {
                        // Ignore
                    }

                    long id = pair.getFirst();
                    return remoteDataSource.getPrice(id).map(priceDto -> new Price(priceDto.getCheeseId(), priceDto.getPrice()));
                });
    }
}

File: app/src/main/java/com/vikingsen/cheesedemo/model/data/price/PriceRepository.kt
package com.vikingsen.cheesedemo.model.data.price

import android.arch.lifecycle.LiveData
import android.arch.lifecycle.Transformations
import com.vikingsen.cheesedemo.model.Resource
import com.vikingsen.cheesedemo.model.webservice.dto.PriceDto
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class PriceRepository @Inject
internal constructor(private val remoteDataSource: PriceRemoteDataSource) {

    fun getPrice(cheeseId: Long, @Suppress("UNUSED_PARAMETER") forceRefresh: Boolean): LiveData<Resource<Price>> {
        return Transformations.map(remoteDataSource.getPrice(cheeseId)) {
            @Suppress("RemoveExplicitTypeArguments") //Technically I only need one <Price>
            when {
                it == null -> Resource.Loading<Price>()
                it.successful -> Resource.Success(mapToPrice(it.data))
                else -> Resource.Error<Price>()
            }
        }
    }

    private fun mapToPrice(data: PriceDto?): Price? {
        return data?.let { (_, cheeseId, price) ->
            Price(cheeseId, price)
        }
    }
}
